hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+                    } else {
+                        // Divergent control flow.
+                        todo!();
+                    }
+                }
+                SuccessorKind::Conditional {
+                    taken_target,
+                    not_taken_target,
+                } => {
+                    // If we don't have any TNT choices buffered, get more.
+                    if self.tnts.is_empty() {
+                        if self.seek_tnt()? == false {
+                            // The packet stream was exhausted.
+                            return Ok(None);
+                        }
+                    }
+                    // Find where to go next based on whether the trace took this branch or not.
+                    let target_off = if self.tnts.pop_front().unwrap() {
+                        *taken_target
+                    } else {
+                        if let Some(ntt) = not_taken_target {
+                            *ntt
+                        } else {
+                            // Divergent control flow.
+                            panic!();
+                        }
+                    };
+                    self.cur_loc = ObjLoc::MainObj(target_off);
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
+                }
+                SuccessorKind::Dynamic => {
+                    // We can only know the successor via a TIP update in a packet. Fall through to
+                    // `self.seek_tip()`.
+                    self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                }
+            }
+        }
+
+        // If we get here then we can't statically figure out where to go next. A packet is going
+        // to have to tell us what to do.
+        self.seek_tip()
+    }
+
+    /// Keep decoding packets until we have some TNT decisions buffered.
+    fn seek_tnt(&mut self) -> Result<bool, HWTracerError> {
+        while self.tnts.is_empty() {
+            let pkt = self.packet()?; // Potentially populates `self.tnts`.
+            if pkt.is_none() {
+                return Ok(false); // Packet stream exhausted.
+            }
+        }
+        Ok(true)
+    }
+
+    /// Keep decoding packets until we receive a TIP update.
+    fn seek_tip(&mut self) -> Result<Option<Block>, HWTracerError> {
+        debug_assert_eq!(self.cur_loc, ObjLoc::OtherObjOrUnknown);
+        loop {
+            if let Some(_) = self.packet()? {
+                if let ObjLoc::MainObj(off) = self.cur_loc {
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(off)));
+                }
+            } else {
+                return Ok(None); // No more packets.
+            }
+        }
+    }
+
+    /// Fetch the next packet and update iterator state.
+    fn packet(&mut self) -> Result<Option<Packet>, HWTracerError> {
+        let ret = if let Some(pkt_or_err) = self.parser.next() {
```

Comment:
Te `let ret = ` is unnecessary. Remove that and you can also remove the `ret` at the end of the function.

Reply:
Debugging left-over :)

Reply:
Fixed in f115368476ec9f8a99db8e4d4467efe9af7c235d